google/cloud/spanner/timestamp.h

```diff
@@ -16,26 +16,137 @@
 #define GOOGLE_CLOUD_CPP_SPANNER_GOOGLE_CLOUD_SPANNER_TIMESTAMP_H_
 
 #include "google/cloud/spanner/version.h"
+#include "google/cloud/status_or.h"
+#include <google/protobuf/timestamp.pb.h>
 #include <chrono>
+#include <cstdint>
+#include <limits>
+#include <ostream>
+#include <string>
+#include <tuple>
 
 namespace google {
 namespace cloud {
 namespace spanner {
 inline namespace SPANNER_CLIENT_NS {
 
 /**
- * An instant in time. A std::chrono::time_point using the system (walltime)
- * clock, represented as a signed integer count (>= 64 bits) of nanoseconds
- * from the clock's epoch.
+ * A representation of the Spanner TIMESTAMP type: An instant in time.
  *
- * Most implementations use an epoch of 1970-01-01T00:00:00+00:00, where the
- * range of values would be about 1677-09-21 through to 2262-04-11.
+ * A `Timestamp` represents an absolute point in time (i.e., is independent of
+ * any time zone), with at least nanosecond precision, and with a range of at
+ * least 0001-01-01T00:00:00Z to 9999-12-31T23:59:59.999999999Z.
  *
- * Note that this may differ from std::chrono::system_clock::time_point on a
- * particular platform.
+ * Note that the string and protobuf representations share those precision and
+ * range guarantees.
+ *
+ * The `std::chrono::time_point` conversions depend on the precision/range of
+ * the underlying duration and on the `std::chrono::system_clock` epoch.
  */
-using Timestamp = std::chrono::time_point<std::chrono::system_clock,
-                                          std::chrono::nanoseconds>;
+class Timestamp {
+ private:
+  // Convenience alias. `std::chrono::sys_time` since C++20.
+  template <class Duration>
+  using sys_time = std::chrono::time_point<std::chrono::system_clock, Duration>;
+
+ public:
+  /// Default construction yields the Unix epoch.
+  Timestamp() : Timestamp(0, 0) {}
+
+  /// Conversion from/to RFC3339 string.
+  ///@{
+  static StatusOr<Timestamp> FromString(std::string const&);
+  std::string ToString() const;
+  ///@}
+
+  /// Conversion from/to `google::protobuf::Timestamp`. These conversions never
+  /// fail, but may accept/produce protobufs outside their documented range.
+  ///@{
+  static Timestamp FromProto(google::protobuf::Timestamp const&);
+  google::protobuf::Timestamp ToProto() const;
+  ///@}
+
+  /// Conversion from/to `std::chrono::time_point` on the system clock.
+  /// May produce out-of-range errors in either direction, depending on the
+  /// properties of `Duration` and the `std::chrono::system_clock` epoch.
+  ///@{
+  template <class Duration>
```

Comment:
Will this allow any `Duration`? If not, if an invalid `Duration` is specified, will an error `Status` be returned? 

If not, we could write a small type trait to verify that the given Durations are valid, then we could static_assert() in the function body.

Reply:
Yes, it does allow any `Duration`.

That said, it is difficult to convert between sec/nsec and a `time_point` that counts in, say, 987654321/1234567890 of a second without risking intermediate overflow or precision loss.  The current implementation only claims to avoid (most of) those problems for the standard durations (see the comments on `FromRatio()` and `ToRatio()`).